File: lib/collect/package.py
import os,re,argparse,glob
import collect

def process_package(context, package_dir, exclude_patterns, expected_use_flags, no_copy = False):
    def is_needed(filename, exclude_patterns):
        exclude_prefixes = ["/usr/share/doc/", "/usr/share/info/", "/usr/share/gtk-doc/", "/usr/share/man/", "/usr/include/", "/usr/lib/pkgconfig/", "/usr/lib64/pkgconfig/","/usr/share/pkgconfig/","/dev/","/etc/portage/"]
        for pfx in exclude_prefixes:
            if filename.startswith(pfx): return False

        for ptn in exclude_patterns + [r"^/usr/lib(64)?/.*lib.+\.a$",r"^/usr/share/locale/(?!ja).+?/"]:
            if re.compile(ptn).match(filename): return False
        return True

    # Check USE flag requirements
    use_flags = open("%s/USE" % package_dir).read().split()
    for u in expected_use_flags:
        if u.startswith('-'):
            if u[1:] in use_flags: raise Exception("%s: USE flag %s is set when it shouldn't be." % (package_dir, u[1:]))
        elif u not in use_flags: raise Exception("%s: USE flag %s is missing when it should be." % (package_dir, u))

    if no_copy: return

    #else
    with open("%s/CONTENTS" % package_dir) as f:
        line = f.readline()
        while line:
            splitted = line.strip().split(' ')
            ent_type = splitted[0] if len(splitted) > 0 else None
            filename = None
            if ent_type == "obj":
                filename = ' '.join(splitted[1:-2])
            elif ent_type == "sym":
                filename = ' '.join(splitted[1:-1]).split(" -> ")[0]

            src = os.path.normpath("%s%s" % (context.source, filename))
            dest = os.path.normpath("%s%s" % (context.destination, filename))

            if filename is not None and is_needed(filename, exclude_patterns) and (os.path.exists(src) or not os.path.exists(dest)):
                collect.process_file(context, filename)
            line = f.readline()

def apply(context, args, dry_run = False):
    parser = argparse.ArgumentParser()
    parser.add_argument("--exclude", action="append", default=[], help="exclude pattern(can be used multiple times)")
    parser.add_argument("--use", type=str, default="", help="USE flags to be expected")
    parser.add_argument("--no-copy", action="store_true", help="just make sure that package exists")
    parser.add_argument("--optional", action="store_true", help="just ignore if not exist")
    parser.add_argument("package", type=str, help="package name")
    args = parser.parse_args(args)
    target_package = context.apply_variables(args.package)
    match = glob.glob(os.path.normpath("%s/var/db/pkg/%s/CONTENTS" % (context.source, target_package)))
    if len(match) < 1:
        match = glob.glob(os.path.normpath("%s/var/db/pkg/%s-[0-9]*/CONTENTS" % (context.source, target_package)))
    if len(match) < 1:
        if args.optional:
            print "Package doesn't match: '%s'" % target_package
            return None
        raise Exception("Package doesn't match: '%s'" % target_package)
    elif len(match) > 1:
        raise Exception("Package anbiguous: '%s', %d packages match" % (target_package, len(match)))

    process_package(context, os.path.dirname(match[0]), args.exclude, args.use.split(), args.no_copy or dry_run)
    return match[0]

File: lib/collect/package.ts
import * as path from "path";
import * as glob from "glob";
import * as fs from "fs-extra";
import {Context} from "./context";
import {file} from "./file";

function find_package(context:Context, pkgname:string):string[]
{
  try {
    const pkgpath = path.join(context.srcdir, "var/db/pkg", pkgname);
    if (fs.statSync(pkgpath).isDirectory()) return [pkgpath];
  }
  catch (e) {
    //
  }
  //else
  const category_and_name = pkgname.split('/', 2);
  const category = category_and_name.length > 1? category_and_name.shift() : "*";
  const name = category_and_name[0];
  return glob.sync(path.join(context.srcdir, `var/db/pkg/${category}/${name}-*`)).filter (_=> path.basename(_).replace(/-r[0-9]+$/, "").replace(/-[^-]+$/, "") == name);
}

export function process_package(context:Context, pkgname:string, options?:{use?:string, no_elf_cache?:boolean, exclude?:string, no_copy?:boolean})
{
  const matched = find_package(context, pkgname).map (_=> {
    const category = fs.readFileSync(path.join(_, "CATEGORY")).toString().trim();
    const pf = fs.readFileSync(path.join(_, "PF"), "utf-8").trim();
    return {
      name: `${category}/${pf}`,
      category: category,
      pf: pf,
      contents: fs.readFileSync(path.join(_, "CONTENTS"), "utf-8").split("\n")
      .filter( _ => _.indexOf("obj ") === 0 || _.indexOf("sym ") === 0 || _.indexOf("dir ") === 0)
      .map(_ =>
        _.replace(/^(obj\s.+)\s[0-9a-z]+\s\d+$/, "$1")
        .replace(/^(sym\s.+)\s->\s.+$/, "$1")
        .replace(/^.+?\s(.+)$/, "$1")),
      use: fs.readFileSync(path.join(_, "USE"), "utf-8").replace(/(\r\n|\n|\r)/gm, "").split(" ")
    }
  });

  if (matched.length == 0) {
    console.log(`No packages such as "${pkgname}" found.`);
    process.exit(-1);
  }

  if (matched.length > 1) {
    console.log(`Package specification "${pkgname}" is ambigious. Candidates:`);
    matched.forEach(_ => {
      console.log(`${_.name}`);
    });
    process.exit(-1);
  }

  const pkg = matched[0];

  if (context.packages.has(pkg.name)) return; // already collected

  // else
  console.log(`package ${pkg.name}`);
  if (options?.use && options.use.split(" ").some(_ => _.startsWith('-')? pkg.use.includes(_.replace(/^-/,"")) : !pkg.use.includes(_))) {
    console.log(`Package "${pkg.name}" does not satisfy use flag condition "${options.use}"`);
    process.exit(-1);
  }

  context.packages.add(pkg.name);

  if (options?.no_copy) return; // just make sure package is installed

  const excluded_prefixes = [
    "/run",
    "/usr/share/man",
    "/usr/share/doc",
    "/usr/share/info",
    "/usr/include"
  ];

  const included_locales = [
    "ja"
  ]

  pkg.contents.forEach(filename => {
    if (!excluded_prefixes.some(_ =>  _ === filename || filename.startsWith(_ + '/') )
      && (!filename.startsWith("/usr/share/locale/") || included_locales.some( _ => filename.startsWith("/usr/share/locale/" + _ + '/')))
      && (!options.exclude || !filename.match(options.exclude))) {
      file(context, filename, options?.no_elf_cache);
    }
  });
}
